Use New Router

driusan opened the pull request:
This updates LORIS to use the PSR15 RequestHandler/Middleware based router that was implemented in PRs #3477, #3480, and #3513.

Most of the mod_rewrite rules, main.php, and NDB_Caller are replaced by a new index.php, which generates a PSR7 ServerRequestInterface and delegates to the module's Module handler.

AjaxHelper.php remains, but is deprecated, as existing scripts will need to be updated to be proper endpoints in the module's router. (Once this is done, the separate router.php for PHP's built in web server can also be removed, and mod_rewrite dependency for LORIS can be removed.)

main.php and NDB_Caller also remain for now in order to load instruments. Once an instrument module is implemented to handle the instruments routers, they can be removed.

Note that instrument URLs also change from $LORIS/instrumentname to $LORIS/instruments/instrumentname as part of this PR.

xlecours commented on docs/config/apache2-site:
```diff
@@ -16,7 +16,7 @@
 	php_value include_path .:/usr/share/php:%LORISROOT%/project/libraries:%LORISROOT%/php/libraries
 
-	DirectoryIndex main.php index.html
+	#DirectoryIndex main.php index.html
```
is the default index.php?

driusan replied:
The default is "index.html", but the last rewrite rule of "RewriteRule ^(.*)$" catches everything.

xlecours commented on htdocs/main.php:
```diff
@@ -224,7 +224,7 @@ function tplFromRequest($param)
     // Set dependencies if they are not set
     if (!isset($tpl_data['jsfiles']) || !isset($tpl_data['cssfiles'])) {
-        $page = (new NDB_Page(new Module(""), '', '', '', ''));
+        $page = (new NDB_Page(new Module("", ""), '', '', '', ''));
```
quotes uniformity.

xlecours commented on modules/instrument_list/php/instrument_list.class.inc:
```diff
@@ -30,6 +33,9 @@ namespace LORIS\instrument_list;
  */
 class Instrument_List extends \NDB_Menu_Filter
 {
+    protected $Candidate;
```
lowercase`c` ?

xlecours commented on modules/instrument_list/php/instrument_list.class.inc:
```diff
@@ -30,6 +33,9 @@ namespace LORIS\instrument_list;
 class Instrument_List extends \NDB_Menu_Filter
 {
+    protected $Candidate;
+    protected $TimePoint;
```
lowercase `t` ?

xlecours commented on modules/timepoint_list/php/timepoint_list.class.inc:
```diff
@@ -25,6 +27,7 @@ namespace LORIS\timepoint_list;
  */
 class Timepoint_List extends \NDB_Menu
 {
+    protected $CandID;
```
lowercaswe `c` ?

xlecours commented on php/libraries/Candidate.class.inc:
```diff
@@ -52,8 +52,9 @@ class Candidate
+     * @static
      */
-    static function &singleton($candID)
+    static public function &singleton($candID)
```
type hint : Candidate ?

driusan replied:
This was just from me trying to debug something and probably shouldn't be in this PR at all.

Should I remove it? It's still a small step forward if I leave it.

xlecours replied:
No, leave it there. There is no harm in adding public. This function, if things goes as planned, should be removed in the next 10 . years :)

xlecours commented on php/libraries/NDB_BVL_Feedback.class.inc:
```diff
@@ -224,7 +224,7 @@ class NDB_BVL_Feedback
      * @return array
      */
-    function getFeedbackTypes()
+    static function getFeedbackTypes()
```
public
and `php/libraries/BVL_Feedback_Panel.class.inc` is making a static and a non static call

xlecours commented on src/Router/ModuleFileRouter.php:
```diff
@@ -81,12 +83,17 @@ public function handle(ServerRequestInterface $request) : ResponseInterface
         return (new \Zend\Diactoros\Response())
             ->withStatus(404)
-            ->withBody(new StringStream($fullpath . ": File not found"));
+            ->withBody(new StringStream($fullpath . ": File not found"))
```
To use the template, should be :
`return (new \LORIS\Http\Error($request, 404, $fullpath . ": File not found"));`

PapillonMcGill commented:
tried to execute but get the following error
PHP Fatal error:  Uncaught ConfigurationException: Config setting issue_tracker_url does not exist in database
I ensure that all SQL patch from 19 and 19.1 were installed and there is no SQL files in this or references PRs

driusan commented:
@PapillonMcGill ? Did you make that comment on the right PR?

driusan commented:
This is going to the major branch.. you're missing SQL/New_patches/2017-04-24-UseOurIssueTrackerPl0x.sql, which is unrelated to this PR

PapillonMcGill commented:
@driusan Effectively, working now! will proceed with testing

driusan commented:
@PapillonMcGill Thanks for catching the data team helper issue, I'll have to update those links (the URL for instruments is now /instruments/testname, I missed that module..) and I'll have to think about how to handle reliability, since it's in a weird semi-deprecated state right now..

xlecours commented on src/Middleware/UserPageDecorationMiddleware.php:
```diff
@@ -79,14 +80,15 @@ public function process(ServerRequestInterface $request, RequestHandlerInterface
-            $tpl_data['bvl_feedback'] = NDB_BVL_Feedback::bvlFeedbackPossible(
-                $this->page
+            $tpl_data['bvl_feedback'] = \NDB_BVL_Feedback::bvlFeedbackPossible(
+                $page
```
should be `$this->PageName`
(will solve `1) InstrumentListTestIntegrationTest::testPageUIs`)

xlecours commented on modules/timepoint_list/php/timepoint_list.class.inc:
```diff
@@ -66,6 +71,32 @@ class Timepoint_List extends \NDB_Menu
+        ServerRequestInterface $request,
+        RequestHandlerInterface $handler
+    ) : ResponseInterface {
+        $attribute = $request->getAttribute("CandID");
```
`CandID` --> `candID` 
(will solve `2) TimepointListIntegrationTest::testTimepointFilteringRules`)
Why use capital letters at all... just saying

driusan commented:
@PapillonMcGill I think I've addressed the issues with the data team helper module, but I'm not familiar enough with the module and the test plan doesn't have enough detail to know if it's working properly (but the links are definitely working better than before.)

@xlecours @ridz1208 This is now passing Travis (Finally!), so it can be reviewed.

PapillonMcGill commented:
@driusan Datahelper now report that "This page (pagename) is under construction"
while loading the reliability page cause a hole bunch of "undefined offset" and 3 php warning of:
` Cannot modify header information - headers already sent by (output started at /var/www/Loris/modules/reliability/php/reliability.class.inc:735) in /var/www/Loris/htdocs/index.php on line 48, referer: https://*instance*.loris.ca/reliability/`

driusan commented:
@PapillonMcGill Neither the test plan nor the code for the `data_team_helper` module make any reference to reliability. Where are you talking about?

PapillonMcGill commented:
@driusan 
For reliability, path is: clinical -> reliability.
on the page itself, the data table is empty with an "error loading data" box, the warning are from apache log.

driusan commented:
@PapillonMcGill I don't think it's worth spending a lot of time on reliability, because it's a one project module that just can't be removed from LORIS because it can't be fully converted to a standalone module (yet, due to dependencies on NDB_Caller), but I managed to get one reliability instrument to load with the latest commit, it just falls back on main.php (like instruments for now) until it's properly converted.

xlecours commented on php/libraries/NDB_BVL_Feedback.class.inc:
```diff
@@ -1065,25 +1065,22 @@ class NDB_BVL_Feedback
+                return true;
+            } else {
+                $Factory       = NDB_Factory::singleton();
+                $DB            = $Factory->Database();
```
the database function is lowercase

xlecours commented:
@driusan, I will wait for the Travis fixes to get in.
